### PR TITLE
CORE-20534: Add additional VNodeCreatorRole permissions

### PR DESCRIPTION
**Pull request by BenYip123**
- Add additional permissions to VNodeCreatorRole that include onboarding of the vNode and use of BYODB operations
- VNodeCreatorRole now has 32 permissions in total
- Update ClusterBuilder to use a VNodeCreatorUser called `vNodeCreatorClient` that has VNodeCreatorRole by default for most vNode related operations

List of 25 permissions added to VNodeCreatorRole:
```
- GET:/api/$VERSION_PATH_REGEX/virtualnode/status/$VNODE_STATUS_REQ_REGEX
- PUT:/api/$VERSION_PATH_REGEX/virtualnode/$VNODE_SHORT_HASH_REGEX/cpi/$CPI_FILE_CHECKSUM_REGEX
- PUT:/api/$VERSION_PATH_REGEX/virtualnode/$VNODE_SHORT_HASH_REGEX/db
- GET:/api/$VERSION_PATH_REGEX/virtualnode/create/db/crypto
- GET:/api/$VERSION_PATH_REGEX/virtualnode/create/db/uniqueness
- GET:/api/$VERSION_PATH_REGEX/virtualnode/create/db/vault/$CPI_FILE_CHECKSUM_REGEX
- GET:/api/$VERSION_PATH_REGEX/virtualnode/$VNODE_SHORT_HASH_REGEX/db/vault/$CPI_FILE_CHECKSUM_REGEX
- PUT:/api/$VERSION_PATH_REGEX/certificate/cluster/$CERTIFICATE_USAGE_REGEX
- GET:/api/$VERSION_PATH_REGEX/certificate/cluster/$CERTIFICATE_USAGE_REGEX
- GET:/api/$VERSION_PATH_REGEX/certificate/cluster/$CERTIFICATE_USAGE_REGEX/$ALIAS_REGEX
- POST:/api/$VERSION_PATH_REGEX/certificate/$TENANT_ID_REGEX/$KEY_ID_REGEX
- GET:/api/$VERSION_PATH_REGEX/hsm/$TENANT_ID_REGEX/$HSM_CATEGORY_REGEX
- POST:/api/$VERSION_PATH_REGEX/hsm/soft/$TENANT_ID_REGEX/$HSM_CATEGORY_REGEX
- POST:/api/$VERSION_PATH_REGEX/key/$TENANT_ID_REGEX/alias/$ALIAS_REGEX/category/$HSM_CATEGORY_REGEX/scheme/$KEY_SCHEME_REGEX
- GET:/api/$VERSION_PATH_REGEX/key/$TENANT_ID_REGEX
- GET:/api/$VERSION_PATH_REGEX/key/$TENANT_ID_REGEX/schemes/$HSM_CATEGORY_REGEX
- GET:/api/$VERSION_PATH_REGEX/key/$TENANT_ID_REGEX/$KEY_ID_REGEX
- POST:/api/$VERSION_PATH_REGEX/membership/$VNODE_SHORT_HASH_REGEX
- GET:/api/$VERSION_PATH_REGEX/membership/$VNODE_SHORT_HASH_REGEX/$UUID_REGEX
- GET:/api/$VERSION_PATH_REGEX/members/$VNODE_SHORT_HASH_REGEX
- GET:/api/$VERSION_PATH_REGEX/mgm/$VNODE_SHORT_HASH_REGEX/info
- PUT:/api/$VERSION_PATH_REGEX/network/setup/$VNODE_SHORT_HASH_REGEX
- GET:/api/$VERSION_PATH_REGEX/flow/$VNODE_SHORT_HASH_REGEX
- GET:/api/$VERSION_PATH_REGEX/flow/$VNODE_SHORT_HASH_REGEX/?status=$FLOW_STATE_REGEX
- GET:/api/$VERSION_PATH_REGEX/flow/$VNODE_SHORT_HASH_REGEX/$CLIENT_REQ_REGEX
```
7 permissions that vNodeCreatorRole already has:
```
- GET:/api/$VERSION_PATH_REGEX/cpi
- POST:/api/$VERSION_PATH_REGEX/cpi
- GET:/api/$VERSION_PATH_REGEX/cpi/status/$UUID_REGEX
- POST:/api/$VERSION_PATH_REGEX/virtualnode
- GET:/api/$VERSION_PATH_REGEX/virtualnode
- GET:/api/$VERSION_PATH_REGEX/virtualnode/$VNODE_SHORT_HASH_REGEX
- PUT:/api/$VERSION_PATH_REGEX/virtualnode/$VNODE_SHORT_HASH_REGEX/state/$VNODE_STATE_REGEX
```

E2E test changes: [CORE-20534: Test VNodeCreatorRole changes #640](https://github.com/corda/corda-e2e-tests/pull/640)
QA-regpack-test changes: [CORE-20534: Fix VNodeCreatorRoleTests that now have additional permissions #1140](https://github.com/corda/qa-regpack-tests/pull/1140)

### CODE REVIEW

File: testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/ClusterBuilder.kt [hunk 1]
```diff
@@ -25,13 +27,57 @@ class ClusterBuilder(clusterInfo: ClusterInfo, val REST_API_VERSION_PATH: String
         init {
             configureTracing("E2eClusterTracing", null, null, emptyMap())
         }
+        private val lock = ReentrantLock()
     }
 
     private val logger = LoggerFactory.getLogger("ClusterBuilder - ${clusterInfo.id}")
+    private val vNodeCreatorName = "vnodecreatoruser"
```
Comment (vkolomeyko) [inline testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/ClusterBuilder.kt]: Minor:
```suggestion
    private const val vNodeCreatorName = "vnodecreatoruser"
```
and we can probably move it inside `companion object`.

File: testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/ClusterBuilder.kt [hunk 2]
```diff
@@ -25,13 +27,57 @@ class ClusterBuilder(clusterInfo: ClusterInfo, val REST_API_VERSION_PATH: String
...
+                    condition { it.code == 201 }
+                }
+
+                assertWithRetry {
+                    command {
+                        assignRoleToUser(vNodeCreatorName, vNodeCreatorRole["id"].textValue())
+                    }
+                    condition { it.code == 200 }
+                }
+                UnirestHttpsClient(clusterInfo.rest.uri, "vnodecreatoruser", "vnodecreatoruser")
```
Comment (vkolomeyko) [inline testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/ClusterBuilder.kt]:
```suggestion
                UnirestHttpsClient(clusterInfo.rest.uri, vNodeCreatorName, vNodeCreatorName)
```

File: testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/ClusterBuilder.kt [hunk 3]
```diff
@@ -227,19 +276,20 @@ class ClusterBuilder(clusterInfo: ClusterInfo, val REST_API_VERSION_PATH: String
...
-        post("/api/$REST_API_VERSION_PATH/maintenance/virtualnode/$virtualNodeShortId/vault-schema/force-resync", "")
+        initialClient.post("/api/$REST_API_VERSION_PATH/maintenance/virtualnode/$virtualNodeShortId/vault-schema/force-resync", "")
 
     /** Return the status for the given request id */
-    fun cpiStatus(id: String) = client.get("/api/$REST_API_VERSION_PATH/cpi/status/$id")
+    fun cpiStatus(id: String) = vNodeCreatorClient.get("/api/$REST_API_VERSION_PATH/cpi/status/$id")
 
     /** List all CPIs in the system */
-    fun cpiList() = client.get("/api/$REST_API_VERSION_PATH/cpi")
+    fun cpiList() = initialClient.get("/api/$REST_API_VERSION_PATH/cpi")
```
Comment (vkolomeyko) [inline testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/ClusterBuilder.kt]: I wonder if this should be done using `vNodeCreatorRole`:
```suggestion
    fun cpiList() = vNodeCreatorClient.get("/api/$REST_API_VERSION_PATH/cpi")
```

**Comment by ShYogesh-R3** [comment]
We might need to handle regression test cases as well which exists in qa-regpack-tests repo. As they would definitely be failing after this PR merge.

**Review by tlawson3 — DISMISSED** [review]
Tooling: From what I understand, the bootstrap process will now include these extra permissions, and the user won't need to perform any additional actions.

**Comment by BenYip123** [comment]
> We might need to handle regression test cases as well which exists in qa-regpack-tests repo. As they would definitely be failing after this PR merge.

Thanks for the heads up. I'll open a branch in qa-regpack-tests now and run some tests on it with this PR before merging

**Review by vkolomeyko — COMMENTED** [review]
Looks fine in principle, few minor comments posted.

**Comment by ShYogesh-R3** [comment]
> > We might need to handle regression test cases as well which exists in qa-regpack-tests repo. As they would definitely be failing after this PR merge.
> 
> Thanks for the heads up. I'll open a branch in qa-regpack-tests now and run some tests on it with this PR before merging

This is the class which you can execute - 
https://github.com/corda/qa-regpack-tests/blob/main/src/test/kotlin/rest_permission_template_tests/rest_permission_template_base/RestPermissionTemplateTestsBase.kt

**Comment by vkolomeyko** [comment]
> > We might need to handle regression test cases as well which exists in qa-regpack-tests repo. As they would definitely be failing after this PR merge.
> 
> Thanks for the heads up. I'll open a branch in qa-regpack-tests now and run some tests on it with this PR before merging

@ShYogesh-R3 - I am not sure if it would break a QA Regpack. This PR is only adding some ALLOW permissions and not removing any of the existing ones. So, if functionality works with even less permissioned user, it should work equally well when more permissions are added.

**Comment by ShYogesh-R3** [comment]
> > > We might need to handle regression test cases as well which exists in qa-regpack-tests repo. As they would definitely be failing after this PR merge.
> > 
> > 
> > Thanks for the heads up. I'll open a branch in qa-regpack-tests now and run some tests on it with this PR before merging
> 
> @ShYogesh-R3 - I am not sure if it would break a QA Regpack. This PR is only adding some ALLOW permissions and not removing any of the existing ones. So, if functionality works with even less permissioned user, it should work equally well when more permissions are added.

Hi @vkolomeyko ,In qa-regpack there are some negative tests as well which are trying to hit some rest endpoints which it doesn't have authorization earlier but now it has. So, they might fail as per my understanding.

**Review by vkolomeyko — APPROVED** [review]
LGTM

**Review by tlawson3 — APPROVED** [review]

**Pull request merged**